### ISSUE TITLE
Building schema failing on updates with the WP GraphqlQL WooCommerce extension

### ISSUE THREAD

**Issue by zamson** [gatsbyjs/gatsby-source-wordpress-experimental#306]
Having a strange issue when running my local development server. While running gatsby develop I'm getting random updates on the Product post type and it always gets stuck on "building schema". Have to trigger a rebuild every time.

⠀
```
info  gatsby-source-wordpress  update product TITLE (#23508)

success extract queries from components - 0.750s
success write out requires - 0.002s
error Missing onError handler for invocation 'building-schema', error was 'Error: Cannot find ObjectTypeComposer with name WpProduct'. Stacktrace was 'Error: Cannot find ObjectTypeComposer with name WpProduct
    at SchemaComposer.getOTC (/Users/andreas/code/example.com/node_modules/graphql-compose/lib/SchemaComposer.js:463:11)
    at forEach (/Users/andreas/code/example.com/node_modules/gatsby/src/schema/infer/index.js:30:37)
    at Array.forEach (<anonymous>)
    at addInferredTypes (/Users/andreas/code/example.com/node_modules/gatsby/src/schema/infer/index.js:27:13)
    at updateSchemaComposer (/Users/andreas/code/example.com/node_modules/gatsby/src/schema/schema.js:142:9)
    at buildSchema (/Users/andreas/code/example.com/node_modules/gatsby/src/schema/schema.js:62:3)
    at build (/Users/andreas/code/example.com/node_modules/gatsby/src/schema/index.js:105:18)
    at buildSchema (/Users/andreas/code/example.com/node_modules/gatsby/src/services/build-schema.ts:19:3)'
success  gatsby-source-wordpress  diff schemas - 3.372s
success  gatsby-source-wordpress  ingest WPGraphQL schema - 3.568s
success createSchemaCustomization - 3.624s
⠀
info  gatsby-source-wordpress  update user john (#3)
⠀
⠀
⠀
info  gatsby-source-wordpress  update product TITLE (#23508)
⠀
⠀
⠠ building schema
```

I'm running the latest  WP GraphqlQL WooCommerce extension 0.6.1 and WP GrahpQL 1.0 with gatsby-source-wordpress-experimental 3.1.3.

**Comment by TylerBarnes** [comment]
Hi @zamson , can you share a reproduction /graphql url as well as the output of `gatsby info`? Thanks!

**Comment by zamson** [comment]
Hey, I'm not sure I can share the graphql endpoint since it's a private project. Below is the gatsby info:

```
System:
    OS: macOS 10.15.7
    CPU: (4) x64 Intel(R) Core(TM) i5-7267U CPU @ 3.10GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 12.19.0 - ~/.nvm/versions/node/v12.19.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v12.19.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 67.0.3396.87
    Firefox: 83.0
    Safari: 14.0
  npmPackages:
    gatsby: ^2.18.8 => 2.26.1
    gatsby-background-image: ^0.9.14 => 0.9.19
    gatsby-background-image-es5: ^0.9.12 => 0.9.19
    gatsby-cli: ^2.12.105 => 2.13.1
    gatsby-image: ^2.2.34 => 2.5.0
    gatsby-plugin-emotion: ^4.1.18 => 4.4.0
    gatsby-plugin-env-variables: ^2.0.0 => 2.0.0
    gatsby-plugin-google-tagmanager: ^2.1.25 => 2.5.0
    gatsby-plugin-layout: ^1.1.21 => 1.4.0
    gatsby-plugin-manifest: ^2.2.31 => 2.6.1
    gatsby-plugin-netlify: ^2.3.13 => 2.5.0
    gatsby-plugin-offline: ^3.0.27 => 3.4.0
    gatsby-plugin-postcss: ^2.1.18 => 2.3.13
    gatsby-plugin-react-helmet: ^3.1.16 => 3.4.0
    gatsby-plugin-react-svg: ^3.0.0 => 3.0.0
    gatsby-plugin-sass: ^2.1.26 => 2.5.1
    gatsby-plugin-sharp: ^2.3.5 => 2.8.0
    gatsby-plugin-sitemap: ^2.2.28 => 2.6.0
    gatsby-source-filesystem: ^2.1.40 => 2.5.0
    gatsby-source-graphql: ^2.8.0 => 2.8.0
    gatsby-source-wordpress-experimental: ^3.1.3 => 3.1.3
    gatsby-transformer-sharp: ^2.3.7 => 2.6.0
  npmGlobalPackages:
    gatsby-cli: 2.13.1
```

As for digging deeper, I also get build errors on Netlify when a variable product has been updated. Maybe it's related:

```
4:34:56 PM: info  gatsby-source-wordpress  A productVariation was updated, but this node type is excluded in plugin options.
4:34:57 PM: error  gatsby-source-wordpress  (undefined)
4:34:57 PM: 	 GraphQL Request must include at least one of those two parameters: "query" or "queryId"
4:34:57 PM:  Error path: undefined
4:34:57 PM:  If you haven't already, try adding define( 'GRAPHQL_DEBUG', true ); to your wp-config.php for more detailed error messages.
info  gatsby-source-wordpress  GraphQL vars: {"id":"cG9zdDoyMzUwNg=="}
4:34:57 PM: error  gatsby-source-wordpress  Error occured while updating a single "productVariation" node.
```

I have not excluded it in the options, just running some default options setup:

```
{
      resolve: `gatsby-source-wordpress-experimental`,
      options: {
        url: process.env.WP_GRAPHQL_URL,
        verbose: true,
        develop: {
          nodeUpdateInterval: 3000,
          hardCacheMediaFiles: true,
        },
        production: {
          hardCacheMediaFiles: false,
        },
        debug: {
          graphql: {
            showQueryOnError: false,
            showQueryVarsOnError: true,
            copyQueryOnError: true,
            panicOnError: true,
            // a critical error is a WPGraphQL query that returns an error and no response data. Currently WPGQL will error if we try to access private posts so if this is false it returns a lot of irrelevant errors.
            onlyReportCriticalErrors: true,
          },
        },
        excludeFieldNames: [`blocksJSON`, `saveContent`],
        type: {
          Post: {
            limit:
              process.env.NODE_ENV === `development`
                ? // Lets just pull 50 posts in development to make it easy on ourselves.
                  50
                : // and we don't actually need more than 5000 in production for this particular site
                  5000,
          },
        },
      },
    },
```

**Comment by TylerBarnes** [comment]
Hi @zamson , please update to the latest versions of WPGatsby, gatsby-source-wordpress-experimental and gatsby core. I believe this is fixed but I can't verify since I don't have access to your site.

Thanks :D

**Comment by ardiewen** [comment]
@TylerBarnes I'm experiencing the exact same error under the following circumstances:

- Local dev environment with `ENABLE_GATSBY_REFRESH_ENDPOINT=true` enabled
- "gatsby": "2.27.4",
- "gatsby-source-wordpress-experimental": "^3.3.0",
- WPGraphQL 1.0
- WP Gatsby 0.6.7
- WPGraphQL WooCommerce 0.6.1

I get this error message when I update a product and have my build hook in WPGatsby set to localhost:8000/__refresh

You have wp-admin access to one of my sanitized backends already via #312, please let me know if you'd like a minimal repo associated with this backend.

**Comment by zamson** [comment]
> Hi @zamson , please update to the latest versions of WPGatsby, gatsby-source-wordpress-experimental and gatsby core. I believe this is fixed but I can't verify since I don't have access to your site.
> 
> Thanks :D

Great, thanks - the cached builds to netlify is now working 👍

**Comment by TylerBarnes** [comment]
Glad to hear that fixed it, thanks @zamson ! I'm going to close this issue but feel free to comment here or re-open the issue if you're still having problems @ardiewen .

Thanks everyone :D

**Comment by ardiewen** [comment]
Hi @TylerBarnes , unfortunately I'm still experiencing this issue.

My builds on Gatsby Cloud are not failing when using incremental builds, though any changes to the product are not updated unless I clear cache and trigger a manual rebuild. This behaviour is replicable in a local dev environment as well, as described below.

I've updated my minimal reproduction here on the `woocommerce branch` to demonstrate this:
https://github.com/ardiewen/gatsby-wordpress-preview-minimal/tree/woocommerce

You still have an admin account on my sanitized test backend, to which I've added:

WooCommerce v4.7.1
WPGraphQL WooCommerce v0.7.2

In addition to existing plugins:

WPGraphQL v1.0
WP Gatsby 0.6.7 (tested with "gatsby-source-wordpress-experimental": "^3.3.1", "gatsby": "^2.28.0")
WP Gatsby v0.7.2 (tested with "gatsby-source-wordpress-experimental": "^4.0.0", "gatsby": "^2.28.0")

### Steps to reproduce:

1. clone my repo and checkout woocommerce branch
2. add .env.development file with
   - GATSBY_GRAPHQL_URL=https://mybackend/graphql
   - ENABLE_GATSBY_REFRESH_ENDPOINT=true
3. `gatsby develop`
4. `ngrok http 8000` and add the url with `/__refresh` to the build hook
5. Update a product using a base post type field (e.g. name, description, etc, no custom taxonomies, ACFs or other plugins)
6. Click update on the product page.
7. See the following in my gatsby develop instance:

```
success  gatsby-source-wordpress  diff schemas - 0.794s
success  gatsby-source-wordpress  ingest WPGraphQL schema - 0.984s
success createSchemaCustomization - 1.067s
success  gatsby-source-wordpress  fetch root fields - 0.517s
⠀
info  gatsby-source-wordpress  update user saponettidev (#1)
⠀
⠀
⠀
info  gatsby-source-wordpress  update product Test Product (#73)
⠀
success  gatsby-source-wordpress  pull updates since last build - 1.641s
success Checking for changed pages - 0.001s
success source and transform nodes - 1.724s
error Missing onError handler for invocation 'building-schema', error was 'Error: Cannot find ObjectTypeComposer with name WpProduct'. Stacktrace was 'Error: Cannot find ObjectTypeComposer with name WpProduct
    at SchemaComposer.getOTC (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/graphql-compose/lib/SchemaComposer.js:463:11)
    at forEach (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/schema/infer/index.js:30:37)
    at Array.forEach (<anonymous>)
    at addInferredTypes (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/schema/infer/index.js:27:13)
    at updateSchemaComposer (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/schema/schema.js:142:9)
    at buildSchema (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/schema/schema.js:62:3)
    at build (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/schema/index.js:105:18)
    at buildSchema (/home/ardiewen/Projects/__Testing/gatsby-starter-default/node_modules/gatsby/src/services/build-schema.ts:19:3)'
⠐ building schema
```

Output of `gatsby info`:

```
  System:
    OS: Linux 4.15 Ubuntu 18.04.2 LTS (Bionic Beaver)
    CPU: (4) x64 Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz
    Shell: 5.4.2 - /usr/bin/zsh
  Binaries:
    Node: 12.19.0 - ~/.nvm/versions/node/v12.19.0/bin/node
    Yarn: 1.22.10 - ~/Projects/__Testing/gatsby-starter-default/node_modules/.bin/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v12.19.0/bin/npm
  Languages:
    Python: 2.7.17 - /usr/bin/python
  npmPackages:
    gatsby: ^2.28.0 => 2.28.0 
    gatsby-image: ^2.5.0 => 2.5.0 
    gatsby-plugin-manifest: ^2.6.1 => 2.6.1 
    gatsby-plugin-offline: ^3.4.0 => 3.4.0 
    gatsby-plugin-react-helmet: ^3.4.0 => 3.4.0 
    gatsby-plugin-sharp: ^2.8.0 => 2.8.0 
    gatsby-source-filesystem: ^2.5.0 => 2.5.0 
    gatsby-source-wordpress-experimental: ^4.0.0 => 4.0.0 
    gatsby-transformer-sharp: ^2.6.0 => 2.6.0 
```

**Comment by ardiewen** [comment]
Hi @TylerBarnes just wanted to follow up on this. Any chance you can confirm the reproduction and reopen this issue? Or should I open a new issue to track this? Let me know, thanks!

**Comment by TylerBarnes** [comment]
Hi @ardiewen , I've been busy with some things but have this on my list. Thanks for the reminder!

**Comment by ardiewen** [comment]
Hey @TylerBarnes, just wanted to follow up on this with some additional information. Looks like there's some additional work to be done on the `wp-graphql-woocommerce` side of things before this can be fixed.

Some additional findings:

1. First thing I noticed was that revisions for products are not automatically turned on. Ref [wp-graphql-woocommerce #421](https://github.com/wp-graphql/wp-graphql-woocommerce/issues/421)

2. After turning on revisions, I tried preview again and got this error:

```
 ERROR 

 gatsby-source-wordpress  (undefined) location: line 1, column: 56
         Unknown argument "asPreview" on field "product" of type "RootQuery".

 Error path: undefined

 If you haven't already, try adding define( 'GRAPHQL_DEBUG', true ); to your wp-config.php for more detailed error messages.


 ERROR #gatsby-source-wordpress-experimental_111002 
```

At this point, I've confirmed the product Root Query via wp-graphql-woocommerce doesn't support the `asPreview` flag. This needs to be resolved and tested before we revisit this issue again on the source plugin. I guess we can close this for now, thanks.
